[joel-foo] iP

joel-foo opened the pull request:
## DukePro

> "Object-oriented programming is an exceptionally bad idea which could only have originated in California.” – Edsger Dijkstra ([source](https://www.oreilly.com/library/view/learning-functional-programming/9781787281394/6ba3dbf2-ecd5-4b1b-a919-2854bad0686b.xhtml#:~:text=Edsger%20W.-,Dijkstra%20says%20OOP%20is%20a%20bad%20idea,only%20have%20originated%20in%20California.%22))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use
All you need to do is,

1. download it from [here](https://nus-cs2103-ay2324s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE!**

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
``` java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

joel-foo committed:
Edit Minion.java

joel-foo committed:
Done Level-1

joel-foo committed:
Done Level-2

joel-foo committed:
Done Level-3

joel-foo committed:
Done Level-4

joel-foo committed:
Done A-TextUiTesting

joel-foo committed:
Done Level-5

joel-foo committed:
Done Level-6

joel-foo committed:
Done Level-6

joel-foo committed:
Edit exceptions

joel-foo committed:
edited

joel-foo committed:
added varargs

joel-foo committed:
done level-7

joel-foo committed:
done level-7

joel-foo committed:
Merge branch 'branch-Level-7'

joel-foo committed:
Done Level-8

joel-foo committed:
Merge branch 'branch-Level-8'

joel-foo committed:
OOP :D

joel-foo committed:
Merge branch 'A-MoreOOP'

joel-foo committed:
Added packages

joel-foo committed:
Merge branch 'branch-A-Packages'

joel-foo committed:
Merge branch 'branch-A-Gradle'

joel-foo committed:
Modified build.gradle

joel-foo committed:
Done A-Junit

joel-foo committed:
Merge branch 'branch-A-Junit'

joel-foo committed:
Done Level-9

joel-foo committed:
Done A-Javadoc

joel-foo committed:
Merge branch 'branch-A-Javadoc'

joel-foo committed:
Merge branch 'branch-Level-9'

joel-foo committed:
Improve codestyle

joel-foo committed:
Fulfill checkstyle

joel-foo committed:
Merge branch 'branch-A-CheckStyle'

joel-foo committed:
Do Level-10

joel-foo committed:
Merge branch 'branch-Level-10'

joel-foo committed:
Add varargs

joel-foo committed:
Merge branch 'branch-A-Varargs'

Weiennn commented on src/main/java/minion/commands/DeadlineCommand.java:
```diff
@@ -0,0 +1,57 @@
+ * Represents a deadline command.
+ */
+public class DeadlineCommand extends Command {
+    public static final String COMMAND_WORD = "deadline";
```
I like how your const is properly in full caps

Weiennn commented on src/main/java/minion/parser/CommandParser.java:
```diff
@@ -0,0 +1,261 @@
+     * @return the task parsed.
+     * @throws MinionException if unable to parse or invalid argument(s) provided.
+     */
+    public static Command parse(String command) throws MinionException {
```
can the method name be more descriptive? Something like parseCommand

Weiennn commented on src/main/java/minion/parser/CommandParser.java:
```diff
@@ -0,0 +1,261 @@
+        // or, something after deadline - but it's just empty space(s)
+        // empty -> no description; non-empty -> still need to check if description is missing.
+        if (arr.length < 2 || arr[1].isEmpty()) {
+            throw new ParserException(Messages.MESSAGE_DEADLINE_DESCRIPTION_ERROR);
```
I like how thoroughly you checked/handled errors

Weiennn commented on src/main/java/minion/parser/CommandParser.java:
```diff
@@ -0,0 +1,261 @@
+        }
+        String from;
+        String to;
+        switch (strs.length) {
```
Good catch, case statement the same indent as the switch statement

xCOLOURx commented on src/main/java/minion/Minion.java:
```diff
@@ -0,0 +1,39 @@
+     * Gets the command result from a string input.
+     */
+    public CommandResult getCommandResult(String input) throws MinionException, IOException {
+        return CommandParser
+                .parse(input)
+                .execute(tasks, storage);
```
I like the line formatting for this part, very clear and easy to follow.

xCOLOURx commented on src/main/java/minion/commands/DeadlineCommand.java:
```diff
@@ -0,0 +1,57 @@
+        tasks.add(deadline);
+        storage.writeToFile(tasks);
+        return new CommandResult(
+            StringFormatter.format(
+                "Got it. I've added this task:",
+                "\t" + deadline.toString(),
+                "Now you have " + tasks.size() + " tasks in the list."
```
I like your line formatting here too!

xCOLOURx commented on src/main/java/minion/commands/Command.java:
```diff
@@ -0,0 +1,21 @@
+ * Represents a command.
+ */
+public abstract class Command {
+    /**
+     * Returns the command result given the tasks, ui, storage.
+     * @param tasks Task list.
+     * @param storage Storage of chatbot.
+     * @throws MinionException if task cannot be executed.
+     * @throws IOException if there is IO error.
+     */
```
Perhaps you could like a line between the description and the params? 
```suggestion
    /**
     * Returns the command result given the tasks, ui, storage.
     *
     * @param tasks Task list.
     * @param storage Storage of chatbot.
     * @throws MinionException if task cannot be executed.
     * @throws IOException if there is IO error.
     */
```

xCOLOURx commented on src/main/java/minion/commands/DeadlineCommand.java:
```diff
@@ -0,0 +1,57 @@
+     */
+    @Override
+    public boolean equals(Object o) {
+        if (o == null || !(o instanceof DeadlineCommand)) {
```
I like your spaces in between the variables here, easy to follow along.

xCOLOURx commented on src/main/java/minion/common/Messages.java:
```diff
@@ -0,0 +1,26 @@
+ * Represents the messages that the chatbot displays to the user.
+ */
+public class Messages {
+    public static final String MESSAGE_GOODBYE = "Bye. Hope to see you again soon!";
+    public static final String MESSAGE_WELCOME = "Hello! I'm Minion!\n" + "What can I do for you?";
+    public static final String MESSAGE_FILE_NOT_FOUND = "Unable to locate the file at /data/tasks.txt. "
+            + "Creating a new one now.";
+    public static final String MESSAGE_MISSING_COMMAND = ":( OOPS!!! I'm sorry, please input a legit command. :-(";
+    public static final String MESSAGE_NO_UNDERSTAND = ":( OOPS!!! I'm sorry, but I don't know what that means :-(";
+    public static final String MESSAGE_TODO_DESCRIPTION_ERROR = ":( OOPS!!! The description of a todo cannot be empty.";
+    public static final String MESSAGE_DEADLINE_DESCRIPTION_ERROR = ":( OOPS!!! The description of a deadline "
+            + "cannot be empty.";
+    public static final String MESSAGE_DEADLINE_BY_ERROR = ":( OOPS!!! The date of a deadline cannot be empty.";
+    public static final String MESSAGE_EVENT_DESCRIPTION_ERROR = ":( OOPS!!! The description of an event cannot"
+            + "be empty.";
+    public static final String MESSAGE_EVENT_FROM_ERROR = ":( OOPS!!! The from date of an event cannot be empty.";
+    public static final String MESSAGE_EVENT_TO_ERROR = ":( OOPS!!! The to date of an event cannot be empty.";
+    public static final String MESSAGE_FAIL_PARSE_DATE = "Unable to parse date. Ensure it is of the form dd/mm/yyyy!";
+    public static final String MESSAGE_FAIL_PARSE_TIME = "Unable to parse time. Ensure it is form HHmm!";
+    public static final String MESSAGE_IO_EXCEPTION = "IO exception occurred. Please try again.";
+    public static final String MESSAGE_INVALID_TASK_INDEX = ":( OOPS!!! Please enter a valid task number.";
+    public static final String MESSAGE_TASK_NOT_FOUND = ":( OOPS!!! Cannot find what you are looking for!";
```
Perhaps you could add a few newline to separate them out into different groups? Like newline after the welcome message or group the errors together.

xCOLOURx commented on src/main/java/minion/data/task/Deadline.java:
```diff
@@ -0,0 +1,70 @@
+ * Represents a deadline task.
+ */
+public class Deadline extends Task {
+    protected String datetime;
```
Perhaps you could write in camelCase?
```suggestion
    protected String dateTime;
```

xCOLOURx commented on src/main/java/minion/parser/CommandParser.java:
```diff
@@ -0,0 +1,261 @@
+        }
+        String[] strs = arr[1].split("/by");
+        String description;
+        switch (strs.length) {
+        // nothing to left and right
+        case 0:
```
I like the switch formatting.

xCOLOURx reviewed (commented):
Your code is well formatted and simple to follow. 👍

joel-foo committed:
Add assertions

There is a lack of assertions in the code

Assertions are useful as they help programmers
check that our assumptions are correct or any
invariants are not violated

Let's add assertions to parts of the code
where verifying such assumptions is important

joel-foo committed:
Merge pull request #1 from joel-foo/branch-A-Assertions

Add assertions

joel-foo committed:
Improve code quality

There are methods in the CommandParser class that have some common
code. There is also inconsistent usage of this references in some
classes.

The common code causes code duplication.

Extracting the common code into another method allows for less code
duplication.

joel-foo committed:
Add streams

Streams are currently missing in the current code.

Streams can be useful in making the code more elegant.

Used a stream implementation for the find method in TaskList class.

joel-foo committed:
Merge pull request #2 from joel-foo/branch-A-CodeQuality

Improve code quality

joel-foo committed:
Merge branch 'master' into branch-A-Streams

joel-foo committed:
Merge pull request #3 from joel-foo/branch-A-Streams

Add streams

joel-foo committed:
Add CI

joel-foo committed:
Enable auto sorting of tasks by type and datetime

joel-foo committed:
Merge branch 'branch-C-Sort'

joel-foo committed:
Fix tests

joel-foo committed:
Add UG

joel-foo committed:
Fix checkstyle

joel-foo committed:
Resolve edge cases

joel-foo committed:
Update UG
